pyuvdata/utils.pyx

```diff
@@ -0,0 +1,306 @@
+# -*- mode: python; coding: utf-8 -*-
+# Copyright (c) 2020 Radio Astronomy Software Group
+# Licensed under the 2-clause BSD License
+
+# distutils: language = c
+# cython: linetrace=True
+# distutils: define_macros=CYTHON_TRACE_NOGIL=1
+# python imports
+import numpy as np
+import warnings
+# cython imports
+cimport numpy
+cimport cython
+from libc.math cimport sin, cos, sqrt, atan2
+
+# in order to not have circular dependencies
+# define transformation parameters here
+# parameters for transforming between xyz & lat/lon/alt
+gps_b = 6356752.31424518
+gps_a = 6378137
+e_squared = 6.69437999014e-3
+e_prime_squared = 6.73949674228e-3
+
+# make c-viewed versions of these variables
+cdef numpy.float64_t _gps_a = gps_a
+cdef numpy.float64_t _gps_b = gps_b
+cdef numpy.float64_t _e2 = e_squared
+cdef numpy.float64_t _ep2 = e_prime_squared
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cpdef tuple baseline_to_antnums(numpy.ndarray[ndim=1, dtype=numpy.int64_t] baseline):
+  cdef unsigned long n = baseline.size
+  cdef numpy.ndarray[ndim=1, dtype=numpy.int64_t] ant1 = np.empty(n, dtype=np.int64)
+  cdef numpy.ndarray[ndim=1, dtype=numpy.int64_t] ant2 = np.empty(n, dtype=np.int64)
+  cdef long _min = baseline.min()
+  cdef int i
+  # make views as c-contiguous arrays of a known dtype
+  # effectivly turns the numpy array into a c-array
+  cdef numpy.int64_t[::1] _a1 = ant1
+  cdef numpy.int64_t[::1] _a2 = ant2
+  cdef numpy.int64_t[::1] _bl = baseline
+
+  with nogil:
```

Comment:
It's my understanding that sometimes it's not worth the overhead of releasing the GIL, but I think the profiling you've done has demonstrated that we don't pay a penalty, even for small-sized arrays. (Btw, thanks for doing the detailed profiling! It's very nice 👍 )

Reply:
We don't have to release it too. :thinking: Up to you. I think I agree that the penalty we pay isn't enough to make even single calls slower than the current implementation but I'm not married to nogil.

Reply:
I think it's fine to keep in (b/c the profiling proves there isn't a significant penalty), just an idle musing I had.